src/org/jgroups/raft/filelog/LogEntryStorage.java

```diff
@@ -167,28 +167,36 @@ private int appendLocked(FileChannel channel, LogEntry[] entries, int index, lon
       }
 
       lastAppended = index - 1;
-      positionCache.invalidate(index);
-      channel.truncate(position);
+      if (positionCache.invalidateFrom(index)) {
+         channel.truncate(position);
+         channel.force(true);
+      } else {
+         // TODO: evaluate if makes sense to truncate upfront in batches
+         channel.force(false);
+      }
       return term;
    }
 
-   private static Header readHeader(FileChannel channel, long position) throws IOException {
-      ByteBuffer data = ByteBuffer.allocate(HEADER_SIZE);
-      channel.read(data, position);
-      data.flip();
-      if (data.remaining() != HEADER_SIZE) {
+   private static Header readHeader(FileChannel channel, long position, IntFunction<ByteBuffer> bufferFactory) throws IOException {
```

Comment:
instead of `IntFunction`, you can use `LogEntryStorage` as a parameter to fetch the `ByteBuffer`. 
If so, the class field `acquireIOBufferWithOperation` could be removed, and the new method in [1] is no longer necessary.

Similar idea for method `org.jgroups.raft.filelog.LogEntryStorage.Header#readLogEntry()`

[1] (line 191)
https://github.com/belaban/jgroups-raft/pull/118/files#diff-55b0a527f9e308774f93bdc6cc0f4ce0680fb84d0243814f9243836478d35a72R191

Reply:
Great, I've missed it: yep let me improve this part :+1 thanks!